consensus/view_change.go

```diff
@@ -1,687 +1,350 @@
-		if len(consensus.m1Payload) == 0 {
-			// TODO(Chao): explain why ReadySignal is sent only in this case but not the other case.
-			// Make sure the newly proposed block have the correct view ID
-			consensus.SetCurBlockViewID(recvMsg.ViewID)
-			go func() {
-				consensus.ReadySignal <- struct{}{}
-			}()
-		} else {
-			consensus.getLogger().Info().
-				Str("From", consensus.phase.String()).
-				Str("To", FBFTCommit.String()).
-				Msg("[OnViewChange] Switching phase")
-			consensus.switchPhase(FBFTCommit, true)
-			copy(consensus.blockHash[:], consensus.m1Payload[:32])
-			aggSig, mask, err := consensus.ReadSignatureBitmapPayload(consensus.m1Payload, 32)
-
-			if err != nil {
-				consensus.getLogger().Error().Err(err).
-					Msg("[onViewChange] ReadSignatureBitmapPayload Fail")
-				return
-			}
+	vc.viewIDBitmap[recvMsg.ViewID].SetKey(senderKey.Bytes, true)
 
-			consensus.aggregatedPrepareSig = aggSig
-			consensus.prepareBitmap = mask
-			// Leader sign and add commit message
-			block := consensus.FBFTLog.GetBlockByHash(consensus.blockHash)
-			if block == nil {
-				consensus.getLogger().Warn().Msg("[onViewChange] failed to get prepared block for self commit")
-				return
-			}
-			commitPayload := signature.ConstructCommitPayload(consensus.ChainReader,
-				block.Epoch(), block.Hash(), block.NumberU64(), block.Header().ViewID().Uint64())
-			for i, key := range consensus.priKey {
-				if err := consensus.commitBitmap.SetKey(key.Pub.Bytes, true); err != nil {
-					consensus.getLogger().Warn().
-						Msgf("[OnViewChange] New Leader commit bitmap set failed for key at index %d", i)
-					continue
-				}
-
-				if _, err := consensus.Decider.SubmitVote(
-					quorum.Commit,
-					key.Pub.Bytes,
-					key.Pri.SignHash(commitPayload),
-					common.BytesToHash(consensus.blockHash[:]),
-					block.NumberU64(),
-					block.Header().ViewID().Uint64(),
-				); err != nil {
-					consensus.getLogger().Warn().Msg("submit vote on viewchange commit failed")
-					return
-				}
+	return M2, nil, nil
+}
 
+// InitPayload do it once when validator received view change message
+// TODO: if I start the view change message, I should init my payload already
+func (vc *ViewChange) InitPayload(
+	fbftlog *FBFTLog,
+	viewID uint64,
+	blockNum uint64,
+	myPubKey string,
+	privKeys multibls.PrivateKeys,
+) error {
+	vc.vcLock.Lock()
+	defer vc.vcLock.Unlock()
+
+	_, ok := vc.nilSigs[viewID][myPubKey]
+
+	if !ok {
+		vc.getLogger().Info().Msg("[InitPayload] add my M2(NIL) type messaage")
+		for i, key := range privKeys {
+			if err := vc.nilBitmap[viewID].SetKey(key.Pub.Bytes, true); err != nil {
+				vc.getLogger().Warn().Int("index", i).Msg("[InitPayload] nilBitmap setkey failed")
+				continue
 			}
+			vc.nilSigs[viewID][key.Pub.Bytes.Hex()] = key.Pri.SignHash(NIL)
 		}
-
-		consensus.SetViewChangingID(recvMsg.ViewID)
-		msgToSend := consensus.constructNewViewMessage(
-			recvMsg.ViewID, newLeaderPriKey,
-		)
-
-		consensus.getLogger().Warn().
```

Comment:
Let's keep this message? I search for this message for view change debug a lot. You can check it to Info()

Reply:
sure.